File: packages/mcp-editor-metadata/src/loader.ts
import { readFile } from "fs/promises";
import { MetadataFileSchema, type MetadataFile } from "./schema.js";

/**
 * Loads and validates a metadata.json file from disk.
 *
 * @param path Absolute or relative path to the metadata file.
 * @returns The parsed and validated metadata.
 * @throws Error with descriptive message if file is missing, malformed, or invalid.
 */
export async function loadMetadata(path: string): Promise<MetadataFile> {
  let raw: string;
  try {
    raw = await readFile(path, "utf-8");
  } catch (err) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT") {
      throw new Error(`Metadata file not found: ${path}`);
    }
    throw new Error(`Failed to read metadata file ${path}: ${(err as Error).message}`);
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    throw new Error(`Failed to parse metadata file ${path}: ${(err as Error).message}`);
  }

  const result = MetadataFileSchema.safeParse(parsed);
  if (!result.success) {
    const firstIssue = result.error.issues[0];
    throw new Error(
      `Metadata file ${path} failed validation: ${firstIssue.path.join(".")} — ${firstIssue.message}`,
    );
  }

  return result.data;
}

File: packages-py/mcp-editor-metadata/src/mcp_editor_metadata/loader.py
"""Load and validate metadata.json from disk."""

from __future__ import annotations

import json
from pathlib import Path

from pydantic import ValidationError

from .schema import MetadataFile


def load_metadata(path: str | Path) -> MetadataFile:
    """
    Load and validate a metadata.json file from disk.

    Raises:
        FileNotFoundError: if the file does not exist.
        ValueError: if the file is not valid JSON or fails schema validation.
    """
    p = Path(path)

    try:
        raw = p.read_text(encoding="utf-8")
    except FileNotFoundError:
        raise FileNotFoundError(f"Metadata file not found: {p}")

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as e:
        raise ValueError(f"Failed to parse metadata file {p}: {e.msg}") from e

    try:
        return MetadataFile.model_validate(parsed)
    except ValidationError as e:
        first = e.errors()[0]
        loc = ".".join(str(x) for x in first["loc"])
        raise ValueError(
            f"Metadata file {p} failed validation: {loc} — {first['msg']}"
        ) from e
